File: src/workflowsuite/kpi/client/time/OleAutomationDateUtil.java
package workflowsuite.kpi.client.time;

import java.time.Instant;

final class OleAutomationDateUtil {

    private OleAutomationDateUtil() {

    }

    private static final long MILLIS_PER_SECOND = 1000;
    private static final long MILLIS_PER_MINUTE = MILLIS_PER_SECOND * 60;
    private static final long MILLIS_PER_HOUR = MILLIS_PER_MINUTE * 60;
    private static final long MILLIS_PER_DAY = MILLIS_PER_HOUR * 24;

    // Number of days in a non-leap year
    private static final long DAYS_PER_YEAR = 365;
    // Number of days in 4 years
    private static final long DAYS_PER_4_YEARS = DAYS_PER_YEAR * 4 + 1;       // 1461
    // Number of days in 100 years
    private static final long DAYS_PER_100_YEARS = DAYS_PER_4_YEARS * 25 - 1;  // 36524
    // Number of days in 400 years
    private static final long DAYS_PER_400_YEARS = DAYS_PER_100_YEARS * 4 + 1; // 146097

    private static final long DAYS_TO_1899 = DAYS_PER_400_YEARS * 4 + DAYS_PER_100_YEARS * 3 - 367;
    private static final long DAYS_TO_1970 =
            DAYS_PER_400_YEARS * 4 + DAYS_PER_100_YEARS * 3 + DAYS_PER_4_YEARS * 17 + DAYS_PER_YEAR; // 719,162

    static Instant fromOADate(double time) {
        //if (time >= 2958466 || time <= -657435)
        //throw
        long millis = (long) (time * MILLIS_PER_DAY + (time >= 0.0 ? 0.5 : -0.5));
        if (millis < 0) {
            millis -= (millis % MILLIS_PER_DAY) * 2;
        }

        millis += (DAYS_TO_1899 - DAYS_TO_1970) * MILLIS_PER_DAY;

        return Instant.ofEpochMilli(millis);
    }
}

File: src/workflowSuite/kpi/client/time/OleAutomationDateUtil.kt
package workflowsuite.kpi.client.time

import java.time.Instant

// Number of 100ns ticks per time unit
private const val TicksPerMillisecond: Long = 10000
private const val TicksPerSecond = TicksPerMillisecond * 1000
private const val TicksPerMinute = TicksPerSecond * 60
private const val TicksPerHour = TicksPerMinute * 60
private const val TicksPerDay = TicksPerHour * 24

private const val MillisPerSecond = 1000
private const val MillisPerMinute = MillisPerSecond * 60
private const val MillisPerHour = MillisPerMinute * 60
private const val MillisPerDay = MillisPerHour * 24

// Number of days in a non-leap year
private const val DaysPerYear = 365
// Number of days in 4 years
private const val DaysPer4Years = DaysPerYear * 4 + 1       // 1461
// Number of days in 100 years
private const val DaysPer100Years = DaysPer4Years * 25 - 1  // 36524
// Number of days in 400 years
private const val DaysPer400Years = DaysPer100Years * 4 + 1 // 146097

private const val DaysTo1899 : Long = (DaysPer400Years * 4 + DaysPer100Years * 3 - 367).toLong()
private const val DaysTo1970 : Long = (DaysPer400Years * 4 + DaysPer100Years * 3 + DaysPer4Years * 17 + DaysPerYear).toLong() // 719,162

fun fromOADate(time: Double): Instant {
    //if (time >= 2958466 || time <= -657435)
    //throw
    var millis = (time * MillisPerDay + if (time >= 0.0) 0.5 else -0.5).toLong()
    if (millis < 0)
        millis -= (millis % MillisPerDay) * 2

    millis += (DaysTo1899 - DaysTo1970) * MillisPerDay

    return Instant.ofEpochMilli(millis)

}


